File: java/p2/src/p2/exec/BasicQuery.java
package p2.exec;

import java.util.List;
import p2.lang.plan.Predicate;
import p2.types.basic.Tuple;
import p2.types.basic.TupleSet;
import p2.types.exception.P2RuntimeException;
import p2.types.operator.Operator;
import p2.types.operator.Projection;
import p2.types.operator.ScanJoin;
import p2.types.table.Aggregation;
import p2.types.table.Table;
import p2.types.table.TableName;

public class BasicQuery extends Query {
	
	private Aggregation aggregation;
	
	private List<Operator> body;

	public BasicQuery(String program, String rule, Boolean isPublic, Boolean isDelete,
					  Predicate event, Predicate head, List<Operator> body) {
		super(program, rule, isPublic, isDelete, event, head);
		this.aggregation = null;
		this.body        = body;
	}
	
	@Override
	public String toString() {
		String query = "Basic Query Rule " + rule() + 
		               ": input " + input().toString();
		if (body != null) {
			for (Operator oper : body) {
				query += " -> " + oper.toString();
			}
		}
		query += " -> " + output().toString();
		return query;
	}

	@Override
	public TupleSet evaluate(TupleSet input) throws P2RuntimeException {
		if (!input.name().equals(input().name())) {
			throw new P2RuntimeException("Query expects input " + input().name() + 
					                     " but got input tuples " + input.name());
		}
		
		TupleSet tuples = new TupleSet(input.name());
		for (Tuple tuple : input) {
			tuple = tuple.clone();
			tuple.schema(input().schema().clone());
			tuples.add(tuple);
		}
		
		for (Operator oper : body) {
			tuples = (TupleSet) oper.evaluate(tuples);
		}
		
		return tuples;
	}

}

File: lincoln/ruby/lib/exec/basic_query.rb
require 'lib/exec/query'
require 'lib/types/exception/dataflow_runtime_exception'
class BasicQuery < Query

  def initialize(context, program, rule, isPublic, isDelete, event, head, body)
    @aggregation = nil
    require 'ruby-debug'; debugger if body.nil?
    @body        = body
    super(context, program, rule, isPublic, isDelete, event, head)
    @context = context
  end

  def to_s
    String   query = "Basic Query Rule " + rule + 
    ": input " + input.to_s
#    require 'ruby-debug'; debugger
    @body.each do |oper| 
      query += " -> " + oper.to_s
    end
    query += " -> " + output().to_s
    return query
  end

  def evaluate(inval)
    print "running query \"#{rule.to_s}(#{inval.name.to_s})\"\n"    
    if (@input.name != inval.name) then
      raise DataflowRuntimeException, "Query expects input " + @input.name.to_s + 
      " but got input tuples " + inval.name.to_s
    end

    tuples = TupleSet.new(@input.name)
    inval.each do |tuple| 
      require 'ruby-debug'; debugger if tuple.size != @input.schema.size
      tuple = tuple.clone
      tuple.schema = @input.schema.clone
      tuples << (tuple)
    end

#    require 'ruby-debug'; debugger if rule == 'q1_rule'
    @body.each do |oper| 
      # PUT IN SOME CHECK HERE THAT OPER SCHEMA VARIABLES NOT REPLICATED!
      # oper.schema.variables.each do |v|
      tuples = oper.evaluate(tuples)
    end
 #   require 'ruby-debug'; debugger if output.name.to_s == 'runtime::insertionQueue' and tuples.size > 0
    print "    produced #{tuples.size.to_s} tups in #{output.name.to_s}:\n       #{tuples.tups.to_s}\n"
    puts if tuples.tups.size > 0 
    return tuples
  end
end
